components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -706,47 +732,231 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
+                disapprovedScopeConsents.removeAll(
+                        Arrays.asList(ArrayUtils.nullToEmpty(oAuthAdminService.getScopeNames())));
+                // Update user consent by merging existing consent.
+                String userId = getUserIdOfAuthenticatedUser(user);
+                String appId = getAppIdFromClientId(params.getClientId());
+                oAuth2ScopeService.updateUserConsentForApplication(userId, appId,
+                        IdentityTenantUtil.getTenantId(user.getTenantDomain()),
+                        null, disapprovedScopeConsents);
+            }
+        } catch (IdentityOAuthAdminException e) {
+            throw new OAuthSystemException(
+                    "Error occurred while removing OIDC scopes from disapproved OAuth scopes.", e);
+        } catch (IdentityOAuth2ScopeException e) {
+            throw new OAuthSystemException(("Error occurred while updating OAuth scope consents."));
+        }
+    }
+
+    /**
+     * Check if the user has already given consent to required OAuth scopes.
+     *
+     * @param user              Authenticated user.
+     * @param oAuth2Parameters  OAuth2 parameters.
+     * @return  True if user has given consent to all the requested OAuth scopes.
+     * @throws IdentityOAuth2ScopeConsentException
+     * @throws IdentityOAuthAdminException
+     */
+    public static boolean isUserAlreadyConsentedForOAuthScopes(AuthenticatedUser user,
+                                                               OAuth2Parameters oAuth2Parameters)
+            throws IdentityOAuth2ScopeException, IdentityOAuthAdminException, OAuthSystemException {
+
+        List<String> scopesToBeConsented = new ArrayList<>(oAuth2Parameters.getScopes());
+        // Remove OIDC scopes.
+        scopesToBeConsented.removeAll(Arrays.asList(ArrayUtils.nullToEmpty(oAuthAdminService.getScopeNames())));
+        String userId = getUserIdOfAuthenticatedUser(user);
+        String appId = getAppIdFromClientId(oAuth2Parameters.getClientId());
+        return oAuth2ScopeService.hasUserProvidedConsentForAllRequestedScopes(userId, appId,
+                IdentityTenantUtil.getTenantId(user.getTenantDomain()), scopesToBeConsented);
+    }
+
+    /**
+     * Store consent given for OAuth scopes by the user for the application.
+     *
+     * @param user                      Authenticated user.
+     * @param params                    OAuth2 parameters.
+     * @param overrideExistingConsent   True to override existing consent, otherwise merge the new consent with
+     *                                  existing consent.
+     * @throws OAuthSystemException
+     */
+    public static void storeOAuthScopeConsent(AuthenticatedUser user, OAuth2Parameters params,
+                                              boolean overrideExistingConsent) throws OAuthSystemException {
+
+        try {
+            Set<String> userApprovedScopesSet = params.getConsentRequiredScopes();
```

Comment:
Shall we add serveral debug logs to identify which path the execution is going

Reply:
Fixed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1548/commits/27ddbc8fed9c4368ffe7a826b857537a6b8ffff0